config/api_client.go

```diff
@@ -0,0 +1,55 @@
+package config
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/CircleCI-Public/circleci-cli/api/graphql"
+	"github.com/CircleCI-Public/circleci-cli/api/rest"
+	"github.com/CircleCI-Public/circleci-cli/settings"
+)
+
+var (
+	compilePath = "compile-config-with-defaults"
+)
+
+type apiClientVersion string
+
+type APIClient interface {
+	CompileConfig(configContent string, orgID string, params Parameters, values Values) (*ConfigResponse, error)
+}
+
+func newAPIClient(config *settings.Config) (APIClient, error) {
+	hostValue := GetCompileHost(config.Host)
+	restClient := rest.NewFromConfig(hostValue, config)
+
+	version, err := detectAPIClientVersion(restClient)
+	if err != nil {
+		return nil, err
+	}
+
+	switch version {
+	case v1_string:
+		return &v1APIClient{graphql.NewClient(config.HTTPClient, config.Host, config.Endpoint, config.Token, config.Debug)}, nil
+	case v2_string:
+		return &v2APIClient{restClient}, nil
+	default:
+		return nil, fmt.Errorf("Unable to recognise your Server's config file API")
+	}
+}
+
+func detectAPIClientVersion(restClient *rest.Client) (apiClientVersion, error) {
```

Comment:
I think the internals of this function could be explained. In a year's time, it's probable that another developer will face this and wonder why v1 results in a 404. 🙃 

Do you think a comment would be justified here?

Reply:
Agree! Thanks